src/Stratis.Features.Unity3dApi/Controllers/Unity3dController.cs

```diff
@@ -0,0 +1,410 @@
+
+        private readonly ChainIndexer chainIndexer;
+
+        private readonly IStakeChain stakeChain;
+
+        /// <summary>Instance logger.</summary>
+        private readonly ILogger logger;
+
+        public Unity3dController(ILoggerFactory loggerFactory, IAddressIndexer addressIndexer,
+            IBlockStore blockStore, IChainState chainState, Network network, ICoinView coinView, WalletController walletController, ChainIndexer chainIndexer, IStakeChain stakeChain)
+        {
+            Guard.NotNull(loggerFactory, nameof(loggerFactory));
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+            this.addressIndexer = Guard.NotNull(addressIndexer, nameof(addressIndexer));
+            this.blockStore = Guard.NotNull(blockStore, nameof(blockStore));
+            this.chainState = Guard.NotNull(chainState, nameof(chainState));
+            this.network = Guard.NotNull(network, nameof(network));
+            this.coinView = Guard.NotNull(coinView, nameof(coinView));
+            this.walletController = Guard.NotNull(walletController, nameof(walletController));
+            this.chainIndexer = Guard.NotNull(chainIndexer, nameof(chainIndexer));
+            this.stakeChain = Guard.NotNull(stakeChain, nameof(stakeChain));
+        }
+
+        /// <summary>
+        /// Gets UTXOs for specified address.
+        /// </summary>
+        /// <param name="address">Address to get UTXOs for.</param>
+        [Route("getutxosforaddress")]
+        [HttpGet]
+        public GetUTXOsResponseModel GetUTXOsForAddress([FromQuery] string address)
+        {
+            VerboseAddressBalancesResult balancesResult = this.addressIndexer.GetAddressIndexerState(new[] {address});
+
+            if (balancesResult.BalancesData == null || balancesResult.BalancesData.Count != 1)
+            {
+                this.logger.LogWarning("No balances found for address {0}, Reason: {1}", address, balancesResult.Reason);
+                return new GetUTXOsResponseModel() {Reason = balancesResult.Reason};
+            }
+
+            BitcoinAddress bitcoinAddress = this.network.CreateBitcoinAddress(address);
+
+            AddressIndexerData addressBalances = balancesResult.BalancesData.First();
+
+            List<AddressBalanceChange> deposits = addressBalances.BalanceChanges.Where(x => x.Deposited).ToList();
+            long totalDeposited = deposits.Sum(x => x.Satoshi);
+            long totalWithdrawn = addressBalances.BalanceChanges.Where(x => !x.Deposited).Sum(x => x.Satoshi);
+
+            long balanceSat = totalDeposited - totalWithdrawn;
+
+            HashSet<uint256> blocksToRequest = new HashSet<uint256>(deposits.Count);
+
+            foreach (AddressBalanceChange deposit in deposits)
+            {
+                int blockHeight = deposit.BalanceChangedHeight;
+                uint256 blockHash = this.chainState.ConsensusTip.GetAncestor(blockHeight).Header.GetHash();
```

Comment:
It might be very slightly more performant to build a hashset of block heights instead, and then iterate through it to build the list of block hashes. Might not be worth changing though

Reply:
fixed. Added guarantee that we won't look for hashes of blocks which hash we've already added